component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java

```diff
@@ -114,40 +254,61 @@ public boolean canHandle(HttpServletRequest request) {
-                || StringUtils.isNotEmpty(request.getParameter(EmailOTPAuthenticatorConstants.CODE))
-                || StringUtils.isNotEmpty(request.getParameter(EmailOTPAuthenticatorConstants.EMAIL_ADDRESS)));
+        return ((StringUtils.isNotEmpty(request.getParameter(RESEND))
+                && StringUtils.isEmpty(request.getParameter(CODE)))
+                || StringUtils.isNotEmpty(request.getParameter(CODE))
+                || StringUtils.isNotEmpty(request.getParameter(EMAIL_ADDRESS))
+                || StringUtils.isNotEmpty(request.getParameter(USER_NAME)));
     }
 
     @Override
-    public AuthenticatorFlowStatus process(HttpServletRequest request,
-                                           HttpServletResponse response,
+    public AuthenticatorFlowStatus process(HttpServletRequest request, HttpServletResponse response,
                                            AuthenticationContext context)
             throws AuthenticationFailedException, LogoutFailedException {
         // if the logout request comes, then no need to go through and complete the flow.
         if (context.isLogoutRequest()) {
             return AuthenticatorFlowStatus.SUCCESS_COMPLETED;
-        } else if (StringUtils.isNotEmpty(request.getParameter(EmailOTPAuthenticatorConstants.EMAIL_ADDRESS))) {
+        } else if (StringUtils.isNotEmpty(request.getParameter(EMAIL_ADDRESS))) {
             // if the request comes with EMAIL ADDRESS, it will go through this flow.
             initiateAuthenticationRequest(request, response, context);
             return AuthenticatorFlowStatus.INCOMPLETE;
-        } else if (StringUtils.isEmpty(request.getParameter(EmailOTPAuthenticatorConstants.CODE)) &&
-                StringUtils.isEmpty(request.getParameter(EmailOTPAuthenticatorConstants.RESEND))) {
+        } else if (StringUtils.isEmpty(request.getParameter(CODE)) && StringUtils.isEmpty(request.getParameter(RESEND)))
+        {
             // if the request comes with code, it will go through this flow.
-            initiateAuthenticationRequest(request, response, context);
+            if (!isIdfInitiatedFromEmailOTP(context)) {
+                if (context.getLastAuthenticatedUser() == null) {
+                    context.setProperty(WITHOUT_AUTHENTICATED_USER, true);
+                    redirectUserToIdf(request, response, context);
+                return AuthenticatorFlowStatus.INCOMPLETE;
+                }
+            }
+            if (context.getLastAuthenticatedUser() == null) {
+                org.wso2.carbon.user.core.common.User user = resolveUser(request, context);
+                setResolvedUserInContext(context, user);
+            }
+
+            try {
+                initiateAuthenticationRequest(request, response, context);
+            } catch (AuthenticationFailedException e) {
+                // The case where user has entered an invalid username
+                log.error(e);
```

Comment:
Can we guarantee that the `AuthenticationFailedException` only happens for the invalid username scenario? In that case, the error log should be meaningfully enough for the developers to troubleshoot

Reply:
Understood. 
No it can't be guaranteed. Will improve the error log. 
Thanks!